dbsnap_verify/__init__.py

```diff
@@ -0,0 +1,176 @@
+)
+
+from time_funcs import (
+    tomorrow_timestamp,
+    now_datetime,
+    timestamp_to_datetime,
+    datetime_to_date_str,
+    three_days_prior,
+)
+
+import boto3
+
+from sys import stdout
+import logging
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+logger.addHandler(logging.StreamHandler(stdout))
+
+
+def wait(state_doc, rds_session):
+    min_timestamp = state_doc["snapshot_minimum_timestamp"]
+    min_datetime = timestamp_to_datetime(min_timestamp)
+    logger.info("Looking for a snapshot of {} older then {}".format(
+        state_doc["database"],
+        datetime_to_date_str(min_datetime)
+    ))
+    snapshot_desc = get_latest_snapshot(rds_session, state_doc["database"])
+    if snapshot_desc["SnapshotCreateTime"] >= min_datetime:
+        # if the latest snapshot is older then the minimum, restore it.
+        restore(state_doc, rds_session)
+    elif min_datetime >= three_days_prior(now_datetime()):
+        # continue wating for a new snapshot, to restore.
+        transition_state(state_doc, "wait")
+        logger.info("Did not find a snapshot of {} older then {}".format(
+            state_doc["database"],
+            datetime_to_date_str(min_datetime)
+        ))
+        logger.info("Going to sleep.")
+    else:
+        logger.warning("Alert! we never found a snapshot of {} older then {}".format(
+            state_doc["database"],
+            datetime_to_date_str(min_datetime)
+        ))
+        alarm("asdfasdfasdkfjnasdf naskdfn aksdjfn")
+
+
+def restore(state_doc, rds_session):
+    tmp_db_description = get_database_description(
+        rds_session, state_doc["tmp_database"]
+    )
+    if tmp_db_description is None:
+        logger.info(
+            "Restoring snapshot of {database} to {tmp_database}".format(
+                **state_doc
+            )
+        )
+        transition_state(state_doc, "restore")
+        sn_ids = state_doc["database_sn_ids"]
+        if isinstance(sn_ids, basestring):
+            sn_ids = sn_ids.split(",")
+        restore_from_latest_snapshot(rds_session, state_doc["database"], sn_ids)
+    elif tmp_db_description["DBInstanceStatus"] == "available":
+        transition_state(state_doc, "modify")
+        modify(state_doc, rds_session)
+
+
+def modify(state_doc, rds_session):
+    logger.info(
+        "Modifying {tmp_database} master password and security groups".format(
+            **state_doc
+        )
+    )
+    sg_ids = state_doc["database_sg_ids"]
+    if isinstance(sg_ids, basestring):
+        sg_ids = sg_ids.split(",")
+    state_doc["tmp_password"] = modify_db_instance_for_verify(
+        rds_session, state_doc["tmp_database"], sg_ids,
+    )
+    transition_state(state_doc, "verify")
+
+
+def verify(state_doc, rds_session):
+    tmp_db_description = get_database_description(
+        rds_session, state_doc["tmp_database"]
+    )
+    tmp_db_status = tmp_db_description["DBInstanceStatus"]
+    tmp_db_event_messages = rds_event_messages(
```

Comment:
Why do you need to do this if you have received the events themselves when the lambda gets triggered?

Reply:
I'm treating the lambda events themselves as a a web hook, this way I don't have to write any special logic for different types of events (which all have different document schemas), instead whenever I get an event I filter it to make sure its for one of my temporary database instances. If it is, I simply treat the event just like the base cloudwatch event, I download the state doc and continue to the next step, whatever that might be.

As a result before we move from modifying to verifying we quickly check the RDS events api for this database instance (by default only asking for events which occured in the last 60s) and then we make sure the credentials were changed.

This is the normal "web hook" pattern where you never trust webhook's payload, but reach back out to verify it yourself.

Reply:
After code review we opted to using the config instead of env vars.